File: Stocker/ProjetoLDS/Stocker/Frontend/StockerMobileUI/stocker_web_ui/lib/models/productStatisticsDto.dart
/// Representa as estatísticas relacionadas a um produto.
///
/// Esta classe é usada para modelar dados estatísticos como o total gasto, a quantidade total adquirida,
/// o número total de compras e a média de gastos por produto.
class ProductStatisticsDto {
  /// Identificador único do produto.
  final int productId;

  /// Valor total gasto com o produto.
  final double totalSpent;

  /// Quantidade total adquirida do produto.
  final double totalQuantity;

  /// Número total de compras envolvendo o produto.
  final int totalPurchases;

  /// Gasto médio por unidade do produto.
  final double averageSpentByProduct;

  /// Construtor da classe [ProductStatisticsDto].
  ///
  /// - [productId]: Identificador único do produto (obrigatório).
  /// - [totalSpent]: Valor total gasto com o produto (obrigatório).
  /// - [totalQuantity]: Quantidade total adquirida do produto (obrigatório).
  /// - [totalPurchases]: Total de compras realizadas do produto (obrigatório).
  /// - [averageSpentByProduct]: Gasto médio por unidade do produto (obrigatório).
  ProductStatisticsDto({
    required this.productId,
    required this.totalSpent,
    required this.totalQuantity,
    required this.totalPurchases,
    required this.averageSpentByProduct,
  });

  /// Converte um mapa JSON numa instância de [ProductStatisticsDto].
  factory ProductStatisticsDto.fromJson(Map<String, dynamic> json) {
    return ProductStatisticsDto(
      productId: json['productId'] as int,
      totalSpent: (json['totalSpent'] as num).toDouble(),
      totalQuantity: (json['totalQuantity'] as num).toDouble(),
      totalPurchases: json['totalPurchases'] as int,
      averageSpentByProduct: (json['averageSpentByProduct'] as num).toDouble(),
    );
  }

  /// Converte uma instância de [ProductStatisticsDto] para um mapa JSON.
  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'totalSpent': totalSpent,
      'totalQuantity': totalQuantity,
      'totalPurchases': totalPurchases,
      'averageSpentByProduct': averageSpentByProduct,
    };
  }
}

File: Stocker/ProjetoLDS/Stocker/Backend/StockerAPI/StockerAPI/Models/Dto/ProductStatisticsDto.cs
namespace StockerAPI.Models.Dto
{
    /// <summary>
    /// Representa as estatísticas relacionadas a um produto, incluindo o total gasto, a quantidade total comprada, o número de compras e a média gasta por produto.
    /// </summary>
    public class ProductStatisticsDto
    {
        /// <summary>
        /// Identificador único do produto.
        /// </summary>
        public int ProductId { get; set; }

        /// <summary>
        /// Total gasto na compra do produto. Representa o valor total de todas as compras desse produto.
        /// </summary>
        public double TotalSpent { get; set; }

        /// <summary>
        /// Quantidade total comprada do produto. Representa o total de unidades adquiridas do produto.
        /// </summary>
        public double TotalQuantity { get; set; }

        /// <summary>
        /// Número total de compras realizadas para o produto.
        /// </summary>
        public int TotalPurchases { get; set; }

        /// <summary>
        /// Média gasta por cada unidade do produto adquirido.
        /// </summary>
        public double AverageSpentByProduct { get; set; }
    }
}
